File: reference/conformance/python/sharenet_conformance/cbor.py
"""ShareNet Canonical CBOR Profile v1 — Python conformance implementation
(work item R1-003, profile defined by R1-002 in the Rust reference core).

Independent reimplementation pinned by the committed vectors; must agree
with the Rust and TypeScript legs byte-for-byte. Error names match the Rust
``DecodeError::name()`` strings exactly for the cross-language diff.
"""

from __future__ import annotations

MAX_DEPTH = 128

I64_MAX = 2**63 - 1

# Value model: int | bytes | str | list | list[tuple[k, v]] | bool | None


class DecodeError(Exception):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code


class EncodeError(Exception):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code


class CborMap(list):
    """A CBOR map: a list of (key, value) tuples.

    A distinct type (vs plain list = array) so EMPTY maps and EMPTY arrays
    stay distinguishable — the value model must round-trip losslessly.
    """


class _Reader:
    __slots__ = ("buf", "pos")

    def __init__(self, buf: bytes):
        self.buf = buf
        self.pos = 0

    def take(self) -> int:
        if self.pos >= len(self.buf):
            raise DecodeError("Truncated", f"truncated at {self.pos}")
        b = self.buf[self.pos]
        self.pos += 1
        return b

    def take_n(self, n: int) -> bytes:
        if n > len(self.buf) - self.pos:
            raise DecodeError("Truncated", f"truncated at {self.pos}")
        s = self.buf[self.pos : self.pos + n]
        self.pos += n
        return s


def _read_arg(r: _Reader, ai: int, at: int) -> int:
    if ai <= 23:
        return ai
    if ai == 24:
        b = r.take()
        if b <= 23:
            raise DecodeError("NonMinimalInteger", f"non-minimal integer at {at}")
        return b
    if ai == 25:
        s = r.take_n(2)
        v = int.from_bytes(s, "big")
        if v <= 0xFF:
            raise DecodeError("NonMinimalInteger", f"non-minimal integer at {at}")
        return v
    if ai == 26:
        s = r.take_n(4)
        v = int.from_bytes(s, "big")
        if v <= 0xFFFF:
            raise DecodeError("NonMinimalInteger", f"non-minimal integer at {at}")
        return v
    if ai == 27:
        s = r.take_n(8)
        v = int.from_bytes(s, "big")
        if v <= 0xFFFFFFFF:
            raise DecodeError("NonMinimalInteger", f"non-minimal integer at {at}")
        return v
    if ai == 31:
        # reached only for major types 0/1/6 (majors 2-5 route to _read_len)
        raise DecodeError("ReservedAdditionalInfo", f"reserved additional info at {at}")
    # 28..=30
    raise DecodeError("ReservedAdditionalInfo", f"reserved additional info at {at}")


def _read_len(r: _Reader, ai: int, at: int) -> int:
    if ai == 31:
        raise DecodeError("IndefiniteLength", f"indefinite length at {at}")
    return _read_arg(r, ai, at)


def _read_value(r: _Reader, depth: int):
    if depth > MAX_DEPTH:
        raise DecodeError("DepthLimitExceeded", f"depth > {MAX_DEPTH} at {r.pos}")
    at = r.pos
    ib = r.take()
    major = ib >> 5
    ai = ib & 0x1F
    if major == 0:
        n = _read_arg(r, ai, at)
        if n > I64_MAX:
            raise DecodeError("IntegerOutOfRange", f"integer out of range at {at}")
        return n
    if major == 1:
        n = _read_arg(r, ai, at)
        if n > I64_MAX:
            raise DecodeError("IntegerOutOfRange", f"integer out of range at {at}")
        return -1 - n
    if major == 2:
        length = _read_len(r, ai, at)
        return bytes(r.take_n(length))
    if major == 3:
        length = _read_len(r, ai, at)
        raw = r.take_n(length)
        try:
            return raw.decode("utf-8")
        except UnicodeDecodeError:
            raise DecodeError("InvalidUtf8", f"invalid UTF-8 at {at}") from None
    if major == 4:
        n = _read_len(r, ai, at)
        items = []
        for _ in range(n):
            items.append(_read_value(r, depth + 1))
        return items
    if major == 5:
        n = _read_len(r, ai, at)
        entries: list[tuple] = []
        prev: bytes | None = None
        for _ in range(n):
            key_start = r.pos
            k = _read_value(r, depth + 1)
            key_end = r.pos
            cur = bytes(r.buf[key_start:key_end])
            if prev is not None:
                if cur == prev:
                    raise DecodeError("DuplicateMapKey", f"duplicate map key at {key_start}")
                if cur < prev:
                    raise DecodeError("UnsortedMapKeys", f"unsorted map keys at {key_start}")
            v = _read_value(r, depth + 1)
            entries.append((k, v))
            prev = cur
        return CborMap(entries)
    if major == 6:
        tag = _read_arg(r, ai, at)
        raise DecodeError("TagNotAllowed", f"tag {tag} at {at}")
    # major 7
    if ai == 20:
        return False
    if ai == 21:
        return True
    if ai == 22:
        return None
    if ai == 23:
        raise DecodeError("UndefinedNotAllowed", f"undefined at {at}")
    if ai == 24:
        v = r.take()
        raise DecodeError("SimpleValueNotAllowed", f"simple value {v} at {at}")
    if ai in (25, 26, 27):
        raise DecodeError("FloatNotAllowed", f"float at {at}")
    if ai == 31:
        raise DecodeError("BreakByteNotAllowed", f"break byte at {at}")
    if 28 <= ai <= 30:
        raise DecodeError("ReservedAdditionalInfo", f"reserved additional info at {at}")
    raise DecodeError("SimpleValueNotAllowed", f"simple value {ai} at {at}")


def decode(data: bytes):
    if len(data) == 0:
        raise DecodeError("EmptyInput", "empty input")
    r = _Reader(data)
    value = _read_value(r, 1)
    if r.pos != len(data):
        raise DecodeError(
            "TrailingBytes", f"trailing bytes at {r.pos} ({len(data) - r.pos} extra)"
        )
    return value


def _write_head(major: int, arg: int, out: bytearray) -> None:
    m = major << 5
    if arg <= 23:
        out.append(m | arg)
    elif arg <= 0xFF:
        out.append(m | 24)
        out.append(arg)
    elif arg <= 0xFFFF:
        out.append(m | 25)
        out.extend(arg.to_bytes(2, "big"))
    elif arg <= 0xFFFFFFFF:
        out.append(m | 26)
        out.extend(arg.to_bytes(4, "big"))
    else:
        out.append(m | 27)
        out.extend(arg.to_bytes(8, "big"))


def _write_value(value, out: bytearray, depth: int) -> None:
    if depth > MAX_DEPTH:
        raise EncodeError("DepthLimitExceeded", f"depth > {MAX_DEPTH}")
    if isinstance(value, bool):
        out.append(0xF5 if value else 0xF4)
        return
    if value is None:
        out.append(0xF6)
        return
    if isinstance(value, int):
        if value >= 0:
            _write_head(0, value, out)
        else:
            _write_head(1, -1 - value, out)
        return
    if isinstance(value, (bytes, bytearray)):
        _write_head(2, len(value), out)
        out.extend(value)
        return
    if isinstance(value, str):
        b = value.encode("utf-8")
        _write_head(3, len(b), out)
        out.extend(b)
        return
    if isinstance(value, CborMap):
        keyed = []
        for i, (k, _) in enumerate(value):
            kb = bytearray()
            _write_value(k, kb, depth + 1)
            keyed.append((bytes(kb), i))
        keyed.sort(key=lambda t: t[0])
        for j in range(1, len(keyed)):
            if keyed[j - 1][0] == keyed[j][0]:
                raise EncodeError("DuplicateMapKey", f"duplicate key at index {keyed[j][1]}")
        _write_head(5, len(value), out)
        for kb, i in keyed:
            out.extend(kb)
            _write_value(value[i][1], out, depth + 1)
        return
    if isinstance(value, list):
        _write_head(4, len(value), out)
        for item in value:
            _write_value(item, out, depth + 1)
        return
    raise EncodeError("UnsupportedValue", f"unsupported value {type(value)}")


def encode(value) -> bytes:
    out = bytearray()
    _write_value(value, out, 1)
    return bytes(out)


def value_from_json(j: dict):
    t = j["type"]
    if t == "int":
        return int(j["value"])
    if t == "bytes":
        return bytes.fromhex(j["hex"])
    if t == "text":
        return j["value"]
    if t == "array":
        return [value_from_json(x) for x in j["items"]]
    if t == "map":
        return CborMap(
            (value_from_json(e["key"]), value_from_json(e["value"])) for e in j["entries"]
        )
    if t == "bool":
        return bool(j["value"])
    if t == "null":
        return None
    raise ValueError(f"unknown JSON value type {t}")


def value_eq(a, b) -> bool:
    """Order-insensitive value equality (map entries as multisets)."""
    if isinstance(a, bool) or isinstance(b, bool):
        return isinstance(a, bool) and isinstance(b, bool) and a == b
    if a is None or b is None:
        return a is None and b is None
    if isinstance(a, int) and isinstance(b, int):
        return a == b
    if isinstance(a, (bytes, bytearray)) and isinstance(b, (bytes, bytearray)):
        return bytes(a) == bytes(b)
    if isinstance(a, str) and isinstance(b, str):
        return a == b
    if isinstance(a, list) and isinstance(b, list):
        a_map = a if isinstance(a, CborMap) else None
        b_map = b if isinstance(b, CborMap) else None
        if (a_map is None) != (b_map is None):
            return False
        if a_map is not None and b_map is not None:
            if len(a_map) != len(b_map):
                return False
            b_dict: dict = {}
            for k, v in b_map:
                b_dict.setdefault(_hashable(k), []).append(v)
            for k, v in a_map:
                bucket = b_dict.get(_hashable(k))
                if bucket is None or not any(value_eq(v, x) for x in bucket):
                    return False
            return True
        if len(a) != len(b):
            return False
        return all(value_eq(x, y) for x, y in zip(a, b))
    return False


def _as_map(v: list):
    if v and all(isinstance(e, tuple) and len(e) == 2 for e in v):
        return v
    return None


def _hashable(v):
    if isinstance(v, CborMap):
        return ("map", tuple((_hashable(k), _hashable(x)) for k, x in v))
    if isinstance(v, list):
        return ("arr", tuple(_hashable(x) for x in v))
    if isinstance(v, (bytes, bytearray)):
        return ("b", bytes(v))
    return (type(v).__name__, v)


def to_hex(data: bytes) -> str:
    return data.hex()

File: reference/conformance/typescript/cbor.ts
/**
 * ShareNet Canonical CBOR Profile v1 — TypeScript conformance implementation
 * (work item R1-003, profile defined by R1-002 in the Rust reference core).
 *
 * This is an INDEPENDENT reimplementation of the profile pinned by
 * `reference/crates/sharenet-protocol/tests/vectors/cbor_vectors.json`. It
 * must agree with the Rust core byte-for-byte on every vector; any divergence
 * is a profile break caught by `run_harness.sh`.
 *
 * Profile (normative, identical to the Rust `cbor.rs`):
 * - values: int (i64 range), bytes, UTF-8 text, array, map (unique keys,
 *   canonically sorted), false/true/null;
 * - minimal-length integers and lengths only; definite lengths only;
 * - map keys sorted by bytewise lexicographic order of their canonical
 *   encodings; duplicates forbidden;
 * - forbidden: tags, floats, indefinite lengths, undefined, other simple
 *   values, trailing bytes, non-minimal integers, invalid UTF-8, truncation;
 * - nesting depth capped at 128 in both directions.
 *
 * Error names match the Rust `DecodeError::name()` strings exactly
 * (EmptyInput, Truncated, TrailingBytes, ReservedAdditionalInfo,
 * IndefiniteLength, NonMinimalInteger, UnsortedMapKeys, DuplicateMapKey,
 * IntegerOutOfRange, TagNotAllowed, FloatNotAllowed, UndefinedNotAllowed,
 * SimpleValueNotAllowed, BreakByteNotAllowed, InvalidUtf8,
 * DepthLimitExceeded) so the conformance harness can diff reject outcomes
 * across languages.
 *
 * Integers use JS `bigint` (the profile's i64 range exceeds Number).
 */

export const MAX_DEPTH = 128;

export type Value =
  | { t: "int"; v: bigint }
  | { t: "bytes"; v: Uint8Array }
  | { t: "text"; v: string }
  | { t: "array"; v: Value[] }
  | { t: "map"; v: [Value, Value][] }
  | { t: "bool"; v: boolean }
  | { t: "null" };

const I64_MAX = 9223372036854775807n;

export class DecodeError extends Error {
  readonly code: string;
  constructor(code: string, message: string) {
    super(message);
    this.code = code;
  }
}

export class EncodeError extends Error {
  readonly code: string;
  constructor(code: string, message: string) {
    super(message);
    this.code = code;
  }
}

class Reader {
  pos = 0;
  constructor(readonly buf: Uint8Array) {}

  take(): number {
    if (this.pos >= this.buf.length) {
      throw new DecodeError("Truncated", `truncated at ${this.pos}`);
    }
    return this.buf[this.pos++]!;
  }

  takeN(n: bigint): Uint8Array {
    const avail = BigInt(this.buf.length - this.pos);
    if (n > avail) {
      throw new DecodeError("Truncated", `truncated at ${this.pos}`);
    }
    const len = Number(n);
    const s = this.buf.subarray(this.pos, this.pos + len);
    this.pos += len;
    return s;
  }
}

function readArg(r: Reader, ai: number, at: number): bigint {
  if (ai <= 23) return BigInt(ai);
  if (ai === 24) {
    const b = r.take();
    if (b <= 23) {
      throw new DecodeError("NonMinimalInteger", `non-minimal integer at ${at}`);
    }
    return BigInt(b);
  }
  if (ai === 25) {
    const s = r.takeN(2n);
    const v = (BigInt(s[0]!) << 8n) | BigInt(s[1]!);
    if (v <= 0xffn) {
      throw new DecodeError("NonMinimalInteger", `non-minimal integer at ${at}`);
    }
    return v;
  }
  if (ai === 26) {
    const s = r.takeN(4n);
    let v = 0n;
    for (let i = 0; i < 4; i++) v = (v << 8n) | BigInt(s[i]!);
    if (v <= 0xffffn) {
      throw new DecodeError("NonMinimalInteger", `non-minimal integer at ${at}`);
    }
    return v;
  }
  if (ai === 27) {
    const s = r.takeN(8n);
    let v = 0n;
    for (let i = 0; i < 8; i++) v = (v << 8n) | BigInt(s[i]!);
    if (v <= 0xffffffffn) {
      throw new DecodeError("NonMinimalInteger", `non-minimal integer at ${at}`);
    }
    return v;
  }
  if (ai === 31) {
    // Reached only for major types 0/1/6 (majors 2-5 route to readLen first).
    throw new DecodeError("ReservedAdditionalInfo", `reserved additional info at ${at}`);
  }
  // 28..=30
  throw new DecodeError("ReservedAdditionalInfo", `reserved additional info at ${at}`);
}

function readLen(r: Reader, ai: number, at: number): bigint {
  if (ai === 31) {
    throw new DecodeError("IndefiniteLength", `indefinite length at ${at}`);
  }
  return readArg(r, ai, at);
}

function cmpBytes(a: Uint8Array, b: Uint8Array): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    if (a[i]! !== b[i]!) return a[i]! < b[i]! ? -1 : 1;
  }
  if (a.length !== b.length) return a.length < b.length ? -1 : 1;
  return 0;
}

const utf8Decoder = new TextDecoder("utf-8", { fatal: true });
const utf8Encoder = new TextEncoder();

function readValue(r: Reader, depth: number): Value {
  if (depth > MAX_DEPTH) {
    throw new DecodeError("DepthLimitExceeded", `depth > ${MAX_DEPTH} at ${r.pos}`);
  }
  const at = r.pos;
  const ib = r.take();
  const major = ib >> 5;
  const ai = ib & 0x1f;
  switch (major) {
    case 0: {
      const n = readArg(r, ai, at);
      if (n > I64_MAX) {
        throw new DecodeError("IntegerOutOfRange", `integer out of range at ${at}`);
      }
      return { t: "int", v: n };
    }
    case 1: {
      const n = readArg(r, ai, at);
      if (n > I64_MAX) {
        throw new DecodeError("IntegerOutOfRange", `integer out of range at ${at}`);
      }
      return { t: "int", v: -1n - n };
    }
    case 2: {
      const len = readLen(r, ai, at);
      return { t: "bytes", v: new Uint8Array(r.takeN(len)) };
    }
    case 3: {
      const len = readLen(r, ai, at);
      const bytes = r.takeN(len);
      let text: string;
      try {
        text = utf8Decoder.decode(bytes);
      } catch {
        throw new DecodeError("InvalidUtf8", `invalid UTF-8 at ${at}`);
      }
      return { t: "text", v: text };
    }
    case 4: {
      const n = readLen(r, ai, at);
      const items: Value[] = [];
      for (let i = 0n; i < n; i++) {
        items.push(readValue(r, depth + 1));
      }
      return { t: "array", v: items };
    }
    case 5: {
      const n = readLen(r, ai, at);
      const entries: [Value, Value][] = [];
      let prev: Uint8Array | null = null;
      for (let i = 0n; i < n; i++) {
        const keyStart = r.pos;
        const k = readValue(r, depth + 1);
        const keyEnd = r.pos;
        const cur = r.buf.subarray(keyStart, keyEnd);
        if (prev !== null) {
          const c = cmpBytes(cur, prev);
          if (c === 0) {
            throw new DecodeError("DuplicateMapKey", `duplicate map key at ${keyStart}`);
          }
          if (c < 0) {
            throw new DecodeError("UnsortedMapKeys", `unsorted map keys at ${keyStart}`);
          }
        }
        const v = readValue(r, depth + 1);
        entries.push([k, v]);
        prev = cur;
      }
      return { t: "map", v: entries };
    }
    case 6: {
      const tag = readArg(r, ai, at);
      throw new DecodeError("TagNotAllowed", `tag ${tag} at ${at}`);
    }
    case 7: {
      if (ai === 20) return { t: "bool", v: false };
      if (ai === 21) return { t: "bool", v: true };
      if (ai === 22) return { t: "null" } as Value;
      if (ai === 23) {
        throw new DecodeError("UndefinedNotAllowed", `undefined at ${at}`);
      }
      if (ai === 24) {
        const v = r.take();
        throw new DecodeError("SimpleValueNotAllowed", `simple value ${v} at ${at}`);
      }
      if (ai === 25 || ai === 26 || ai === 27) {
        throw new DecodeError("FloatNotAllowed", `float at ${at}`);
      }
      if (ai === 31) {
        throw new DecodeError("BreakByteNotAllowed", `break byte at ${at}`);
      }
      if (ai >= 28 && ai <= 30) {
        throw new DecodeError("ReservedAdditionalInfo", `reserved additional info at ${at}`);
      }
      throw new DecodeError("SimpleValueNotAllowed", `simple value ${ai} at ${at}`);
    }
    default:
      throw new Error("unreachable: major is 3 bits");
  }
}

export function decode(bytes: Uint8Array): Value {
  if (bytes.length === 0) {
    throw new DecodeError("EmptyInput", "empty input");
  }
  const r = new Reader(bytes);
  const value = readValue(r, 1);
  if (r.pos !== bytes.length) {
    throw new DecodeError(
      "TrailingBytes",
      `trailing bytes at ${r.pos} (${bytes.length - r.pos} extra)`,
    );
  }
  return value;
}

function writeHead(major: number, arg: bigint, out: number[]) {
  const m = major << 5;
  if (arg <= 23n) {
    out.push(m | Number(arg));
  } else if (arg <= 0xffn) {
    out.push(m | 24, Number(arg));
  } else if (arg <= 0xffffn) {
    out.push(m | 25, Number(arg >> 8n), Number(arg & 0xffn));
  } else if (arg <= 0xffffffffn) {
    out.push(m | 26);
    for (let i = 3; i >= 0; i--) out.push(Number((arg >> BigInt(8 * i)) & 0xffn));
  } else {
    out.push(m | 27);
    for (let i = 7; i >= 0; i--) out.push(Number((arg >> BigInt(8 * i)) & 0xffn));
  }
}

function writeValue(value: Value, out: number[], depth: number): void {
  if (depth > MAX_DEPTH) {
    throw new EncodeError("DepthLimitExceeded", `depth > ${MAX_DEPTH}`);
  }
  switch (value.t) {
    case "int":
      if (value.v >= 0n) {
        writeHead(0, value.v, out);
      } else {
        writeHead(1, -1n - value.v, out);
      }
      break;
    case "bytes":
      writeHead(2, BigInt(value.v.length), out);
      for (const b of value.v) out.push(b);
      break;
    case "text": {
      const b = utf8Encoder.encode(value.v);
      writeHead(3, BigInt(b.length), out);
      for (const x of b) out.push(x);
      break;
    }
    case "array":
      writeHead(4, BigInt(value.v.length), out);
      for (const item of value.v) writeValue(item, out, depth + 1);
      break;
    case "map": {
      const keyed = value.v.map(([k], i) => {
        const kb: number[] = [];
        writeValue(k, kb, depth + 1);
        return { kb: Uint8Array.from(kb), i };
      });
      keyed.sort((a, b) => cmpBytes(a.kb, b.kb));
      for (let j = 1; j < keyed.length; j++) {
        if (cmpBytes(keyed[j - 1]!.kb, keyed[j]!.kb) === 0) {
          throw new EncodeError("DuplicateMapKey", `duplicate key at index ${keyed[j]!.i}`);
        }
      }
      writeHead(5, BigInt(value.v.length), out);
      for (const { kb, i } of keyed) {
        for (const x of kb) out.push(x);
        writeValue(value.v[i]![1], out, depth + 1);
      }
      break;
    }
    case "bool":
      out.push(value.v ? 0xf5 : 0xf4);
      break;
    case "null":
      out.push(0xf6);
      break;
  }
}

export function encode(value: Value): Uint8Array {
  const out: number[] = [];
  writeValue(value, out, 1);
  return Uint8Array.from(out);
}

// ---------------------------------------------------------------------------
// JSON value model bridge (vectors store CBOR values as JSON)
// ---------------------------------------------------------------------------

export function valueFromJson(j: any): Value {
  switch (j.type) {
    case "int":
      return { t: "int", v: BigInt(j.value) };
    case "bytes":
      return { t: "bytes", v: fromHex(j.hex) };
    case "text":
      return { t: "text", v: j.value };
    case "array":
      return { t: "array", v: j.items.map(valueFromJson) };
    case "map":
      return {
        t: "map",
        v: j.entries.map(
          (e: any) => [valueFromJson(e.key), valueFromJson(e.value)] as [Value, Value],
        ),
      };
    case "bool":
      return { t: "bool", v: j.value };
    case "null":
      return { t: "null" } as Value;
    default:
      throw new Error(`unknown JSON value type ${j.type}`);
  }
}

export function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

export function toHex(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += b.toString(16).padStart(2, "0");
  return s;
}

export function fromHex(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(2 * i, 2 * i + 2), 16);
  }
  return out;
}
